File: CopasiCS/CPP/Client/ClientNetwork.cpp
/*
 * ClientNetwork.CPP
 *
 */

#include "Message.h"
#include "ClientNetwork.h"
#include "Client.h"


ClientNetwork::ClientNetwork(Client *myClient, string host, int port): reader(this), writer(this)
{

    this->myClient = myClient;
    /*  Create a socket for the client.  */
    sockfd = socket(AF_INET, SOCK_STREAM, 0);
    cout << endl << "Got Socket FileDiscriptor " << sockfd;

    /*  Name the socket, as agreed with the server.  */
    struct sockaddr_in address;

    address.sin_family = AF_INET;
    address.sin_addr.s_addr = inet_addr(host.c_str());
    //address.sin_addr.s_addr = inet_addr("127.0.0.1");
    //address.sin_port = htons(9734);
    address.sin_port = htons(port);

    int len;
    len = sizeof(address);

    /*  Now connect our socket to the server's socket.  */
    int result;
    result = connect(sockfd, (struct sockaddr *)&address, len);

    if(result == -1)
    {
        this->showMessage("Cannot connect to the server.");
        exit(EXIT_FAILURE);
    }

    /*Note that the reader and writer objects have already been initialized
      in the "Initializer List" of the constructor*/
    reader.start();
    //writer.start();
}


int ClientNetwork::getSocketFileDescriptor()
{
    return this->sockfd;
}

void ClientNetwork::sendMessageToServer(Message &msg)
{
    writer.writeMessageToServer(msg);
}

void ClientNetwork::gotMessageFromServer(Message &msg)
{
    showMessage("Inside ClientNetwork::gotMessageFromServer");
    cerr << (void *) myClient;
    this->myClient->gotMessageFromServer(msg);
}

ClientNetwork::~ClientNetwork()
{
    reader.stop();
    //writer.stop();
}

File: CopasiCS/Java/Client/ClientNetwork.java
package Client;

/*
 * ClientNetwork.java
 */

import java.io.*;
import java.net.*;
import Common.*;

/**This class is the Network component used by the server GUI
 * to communicate with the client. It uses sendMessageToClient(Message)
 * to send a message to the client. It calls gotMessageFromClient(Message)
 * of its "Server" (GUI Component) when a message is recieved
 * by its reader
 */
public class ClientNetwork 
{
	
	/**This it the GUI Client of this network component*/
	protected Client myClient;
	
	/**This is the Socket the client will use to communicate with the server*/
	protected Socket socket;
	/**A new thread which is created to read messages sent from the server.
	 * Remains blocked most of the time on a socket read call. Gets unblocked
	 * on recieving a message.
	 */
	protected ClientSocketReader reader; 
	
	/**Writes messages to be sent to the server on the socket.*/
	protected ClientSocketWriter writer;
	
	/**Constructor taking the host to connect to and the port number
	* @param host The name (or IP) of the server to connect to.
	* @param port The port number on which the server is listening.
	*/
	public ClientNetwork(Client myClient, String host, int port) 
	{
		this.myClient = myClient;
		try
		{
			socket = new Socket(host, port);
			reader = new ClientSocketReader(this);
			writer = new ClientSocketWriter(this);
			showMessage("About to start Reader...");
			reader.start();
			//writer.start();
    	}
    	catch (UnknownHostException uhe)
    	{
    		showMessage("The host: " + host + " was not found...Quitting now.");
    		System.exit(1);
    	}
    	catch (IOException e) 
    	{
      		System.err.println(e);
      		System.exit(1);
    	}
	}
	
	
	/**Method to call when the Client needs to send message a message 
	 * to the server.
	 * @param msg The message that is to be sent to the server by the Client.
	 */
	public void sendMessageToServer(Message msg)
	{
		this.writer.writeMessageToServer(msg);		
	}
	
	/**Method that is called by the Reader when it recieves a message
	 * from the server. This method in turn calls the Client (GUI Component) 
	 * version of this method.
	 * @param The message recieved from the server.
	 */
	public void gotMessageFromServer(Message msg)
	{
		this.myClient.gotMessageFromServer(msg);
	}

	
	/**A small function to display messages.
	 * @param msg The message string to show.
	 */
	private static void showMessage(String msg)
	{
		System.err.println("ClientNetwork: " + msg);
	}
	
}
	